File: Estruturais/Decorator/decorator.js
//componente -------------------------------------------------------------------
class ComponenteNotificador {
    enviar(menssagem){
        console.log(`enviando mensagem: ${menssagem}`);
    }
}

//decorator base -------------------------------------------------------------------

class DecoratorNotificador {
    constructor(componente){
        this.componente = componente;
    }

    enviar(mensagem){
        this.componente.enviar(mensagem);
    }
}

//concreto 1 -------------------------------------------------------------------
class DecoratorNotificadorSMS extends DecoratorNotificador {
    enviar(mensagem){
        super.enviar(mensagem);
        console.log(`enviando mensagem por SMS: ${mensagem}`)
    }
}

//concreto 2 -------------------------------------------------------------------
class DecoratorNotificadorFacebook extends DecoratorNotificador {
    enviar(mensagem){
        super.enviar(mensagem);
        console.log(`enviando mensagem por Facebook: ${mensagem}`)
    }
}

//concreto 3 -------------------------------------------------------------------
class DecoratorNotificadorSlack extends DecoratorNotificador {
    enviar(mensagem){
        super.enviar(mensagem);
        console.log(`enviando mensagem por Slack: ${mensagem}`)
    }
}


//cliente -------------------------------------------------------------------
const notificador = new ComponenteNotificador();
const notificadorComSMS = new DecoratorNotificadorSMS(notificador);
const notificadorComFacebook = new DecoratorNotificadorFacebook(notificador);
const notificadorComSlack = new DecoratorNotificadorSlack(notificador);


notificadorComFacebook.enviar("Hoje é sexta graças a Deus não aguento mais")
notificadorComSlack.enviar("preciso de férias")
notificadorComSMS.enviar("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")

File: Estruturais/Decorator/decorator.py
# //componente -------------------------------------------------------------------
class ComponenteNotificador: 
    def enviar(self, mensagem):
        print(f"Enviando menssagem {mensagem}")

# //decorator base -------------------------------------------------------------------

class DecoratorNotificador:
    def __init__(self, componente):
        self.componente = componente

    def enviar(self, mensagem):
        self.componente.enviar(mensagem)     

# //concreto 1 -------------------------------------------------------------------
class DecoratorNotificadorSMS (DecoratorNotificador):
    def enviar(self, mensagem):
        super().enviar(mensagem)
        print(f" enviando mensagem por SMS: {mensagem}") 

# //concreto 2 -------------------------------------------------------------------
class DecoratorNotificadorFacebook (DecoratorNotificador):
    def enviar(self, mensagem):
        super().enviar(mensagem)
        print(f" enviando mensagem por Facebook: {mensagem}") 

# //concreto 3 -------------------------------------------------------------------
class DecoratorNotificadorSlack (DecoratorNotificador):
    def enviar(self, mensagem):
        super().enviar(mensagem)
        print(f" enviando mensagem por Slack: {mensagem}") 


# //cliente -------------------------------------------------------------------
notificador = ComponenteNotificador()
notificadorComSMS = DecoratorNotificadorSMS(notificador)
notificadorComFacebook = DecoratorNotificadorFacebook(notificador)
notificadorComSlack = DecoratorNotificadorSlack(notificador)


notificadorComFacebook.enviar("Hoje é sexta graças a Deus não aguento mais")
notificadorComSlack.enviar("preciso de férias")
notificadorComSMS.enviar("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")
